taichi/codegen/codegen_metal.cpp

```diff
@@ -663,28 +656,52 @@ class KernelCodegen : public IRVisitor {
         (stmt->const_end ? stmt->end_value : stmt->end_offset);
 
     current_appender().push_indent();
+    const std::string total_elems_name("total_elems");
+    // Begin expression of the for range, this can be either a constant
+    // (const_begin == true), or a variable loaded from the global temporaries.
+    std::string begin_expr;
     if (range_for_attribs.const_range()) {
-      ka.num_threads = range_for_attribs.end - range_for_attribs.begin;
+      const int num_elems = range_for_attribs.end - range_for_attribs.begin;
+      begin_expr = std::to_string(stmt->begin_value);
       emit("// range_for, range known at compile time");
-      emit("if ({} >= {}) return;", kKernelThreadIdName, ka.num_threads);
+      emit("const int {} = {};", total_elems_name, num_elems);
+      // We don't clamp this to kMaxNumThreadsGridStrideLoop, because we know
+      // for sure that we need |num_elems| of threads.
+      // sdf_renderer.py benchmark for setting |num_threads|
+      // - num_elemnts: ~20 samples/s
+      // - kMaxNumThreadsGridStrideLoop: ~12 samples/s
+      ka.num_threads = num_elems;
```

Comment:
In order to fit different applications, can we make `kMaxNumThreadsGridStrideLoop` configuable like `ti.cfg.device_memory_GB`?
If so, please use a generic name (i.e. not ti.cfg.metal_xxx, just ti.cfg.xxx) since opengl may implement grid-stride-loop later.

Reply:
Right. Note that `kMaxNumThreadsGridStrideLoop` is currently inside `metal` namespace, so it's not meant to be shared.

Once all the backends adopt this approach, I think we can start adding a new config field as you suggested?